Add Longhorn improvements (#1184)
suse-edge/suse-edge.github.io: asciidoc/guides/longhorn-encryption.adoc

. **Defense in Depth**: Encryption provides an additional security layer that complements Kubernetes RBAC and network policies.
. **Secure Data Disposal**: Encrypted volumes can be securely decommissioned by simply destroying the encryption keys, ensuring data is rendered permanently inaccessible.

[NOTE]
====
Encryption is applied transparently at the volume level. Applications interact with encrypted volumes through standard Kubernetes persistent volume claims without requiring code modifications.
====

== Architecture overview

Longhorn encryption integrates with the Kubernetes Container Storage Interface (CSI) to provide transparent volume encryption. The encryption process occurs during volume provisioning and is managed through the following components:

* **Kubernetes Secrets**: Encryption keys are stored securely as Kubernetes Secret resources
* **StorageClass Parameters**: Encryption configuration is specified through CSI-specific parameters
* **LUKS Device Mapper**: Each encrypted volume is backed by a LUKS-encrypted device mapper target
* **CSI Driver**: The Longhorn CSI driver manages encryption key retrieval and volume staging

> The encryption workflow operates as follows:

Reviewer: Does it have to be numbered or can have bulleted list?

. When a PersistentVolumeClaim (PVC) requests storage from an encrypted StorageClass, the Longhorn CSI driver retrieves the encryption key from the designated Secret
. A LUKS-encrypted device mapper target is created using the encryption key
. All data written to the volume is encrypted before being committed to the underlying storage
. When the volume is mounted, the CSI driver stages the volume by unlocking the LUKS device using the stored encryption key

== Prerequisites

Before implementing Longhorn volume encryption, ensure the following prerequisites are met:

* A functional Kubernetes cluster with Longhorn version 1.11.1 or higher installed
* Cluster administrative access with appropriate RBAC permissions
* `kubectl` command-line tool configured for cluster access
* `openssl` utility for cryptographic key generation

> == Implementation guide

Reviewer: Do we need to call it Implementation Guide? Or can we just retain it as Implementation?

> === Step 1: Generate the encryption key

Reviewer: Any specific reason to have the steps as separate sub sections? 
Can we just have everything numbered under the Implementation section with sub steps?

> Encryption key generation must be performed using cryptographically secure random number generation. Execute the following command to generate a 256-bit encryption key encoded in base64 format:

Reviewer: ```suggestion
Encryption key generation must be performed using cryptographically secure random number generation. Run the following command to generate a 256-bit encryption key encoded in `base64` format:
```

[source,bash]
----
openssl rand -base64 32
----

Example output:
[source,text]
----
Kq8Xh2vN9pL3mZ1wR4tY6uI0oP5aS7dF8gH9jK1lM2n=
----

[WARNING]
====
**Key Management Critical Requirements**

* **Never commit encryption keys to version control systems**
* Loss of the encryption key results in **permanent and irreversible data loss**
====

=== Step 2: Create the encryption key Secret

Once generated, the encryption key must be stored as a Kubernetes Secret in the `longhorn-system` namespace:

[source,bash]
----
CRYPTO_KEY_VALUE="Kq8Xh2vN9pL3mZ1wR4tY6uI0oP5aS7dF8gH9jK1lM2n="

kubectl create secret generic longhorn-crypto \
  --from-literal=CRYPTO_KEY_VALUE="${CRYPTO_KEY_VALUE}" \
  -n longhorn-system
----

Verify successful Secret creation:

[source,bash]
----
kubectl get secret longhorn-crypto -n longhorn-system
----

[TIP]
====

> In production environments, consider using external secret management solutions that integrate with Kubernetes, such as the External Secrets Operator or sealed-secrets, to avoid storing sensitive data directly in cluster etcd.

Reviewer: ```suggestion
In production environments, consider using external secret management solutions that integrate with Kubernetes, such as the External Secrets Operator or sealed-secrets, to avoid storing sensitive data directly in the cluster's etcd.
```

====

=== Step 3: Configure the encrypted StorageClass

> Define a StorageClass that specifies encryption parameters for the Longhorn CSI driver:

Reviewer: ```suggestion
Define a `StorageClass` that specifies encryption parameters for the Longhorn CSI driver:
```

metadata:
  name: longhorn-encrypted
provisioner: driver.longhorn.io
allowVolumeExpansion: true
parameters:
  numberOfReplicas: "3"
  staleReplicaTimeout: "2880"
  fromBackup: ""
  fsType: "ext4"
  dataLocality: "disabled"
  # Encryption configuration
  encrypted: "true"
  csi.storage.k8s.io/provisioner-secret-name: "longhorn-crypto"
  csi.storage.k8s.io/provisioner-secret-namespace: "longhorn-system"
  csi.storage.k8s.io/node-publish-secret-name: "longhorn-crypto"
  csi.storage.k8s.io/node-publish-secret-namespace: "longhorn-system"
  csi.storage.k8s.io/node-stage-secret-name: "longhorn-crypto"
  csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
----

The encryption-specific parameters configure CSI secret references:

* `encrypted: "true"` - Enables LUKS encryption for all volumes provisioned from this StorageClass
* `csi.storage.k8s.io/provisioner-secret-name` - Specifies the Secret containing the encryption key used during volume provisioning
* `csi.storage.k8s.io/node-publish-secret-name` - Identifies the Secret used when mounting volumes to pods
* `csi.storage.k8s.io/node-stage-secret-name` - References the Secret used during volume staging operations

> Apply the StorageClass definition:

Reviewer: All these are sub steps and it is a good practice to use the ordered list.

[source,bash]
----
kubectl apply -f longhorn-encrypted-storageclass.yaml
----

Verify StorageClass creation and configuration:

[source,bash]
----
kubectl get storageclass longhorn-encrypted
kubectl get storageclass longhorn-encrypted -o yaml
----

=== Step 4: Deploy a test workload

To validate the encryption implementation, deploy a test pod that writes data to an encrypted volume.

> Create the following manifest (`test-encryption.yaml`):

Reviewer: These are sub steps needed to deploy a test pod. If we are going to retain this as a separate sub section, then we can have these as numbered steps.

  name: test-encrypted-pvc
  namespace: default
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn-encrypted
  resources:
    requests:
      storage: 1Gi

---
# Pod that writes sensitive data to the volume
apiVersion: v1
kind: Pod
metadata:
  name: encryption-test-pod
  namespace: default
  labels:
    app: encryption-test
spec:
  containers:
  - name: writer
    image: busybox
    command:
      - sh
      - -c
      - |
        echo "=== Writing sensitive data to encrypted volume ==="
        echo "SECRET DATA: Credit Card 4111-1111-1111-1111" > /data/secret.txt
        echo "SECRET DATA: Password MySuperSecret123!" >> /data/secret.txt
        echo "SECRET DATA: API Key sk-1234567890abcdef" >> /data/secret.txt
        date >> /data/secret.txt
        echo "=== Data written successfully ==="
        echo "Content of /data/secret.txt:"
        cat /data/secret.txt
        echo "=== Sleeping forever (use 'kubectl exec' to verify) ==="
        sleep infinity
    volumeMounts:
    - name: encrypted-storage
      mountPath: /data
  volumes:
  - name: encrypted-storage
    persistentVolumeClaim:
      claimName: test-encrypted-pvc
----

This manifest defines:

> * A PersistentVolumeClaim requesting 1Gi of storage from the `longhorn-encrypted` StorageClass

Reviewer: Should this  `PersistentVolumeClaim` be monospaced?

* A Pod that writes test data to the encrypted volume and remains running for verification

Deploy the test workload:

[source,bash]
----
kubectl apply -f test-encryption.yaml
----

Monitor pod startup:

[source,bash]
----
kubectl get pod encryption-test-pod -w
----

=== Step 5: Verify volume functionality

Once the pod reaches `Running` state, verify that data operations complete successfully:

[source,bash]
----
# View pod logs
kubectl logs encryption-test-pod

> # Execute a shell inside the pod

Reviewer: ```suggestion
# Run a shell inside the pod
```

kubectl exec -it encryption-test-pod -- sh

# Inside the pod, verify the content:
cat /data/secret.txt

> ls -lh /data/

Reviewer: Can these be steps instead of being inside the code block?

Thu Jul 03 10:30:45 UTC 2026
----

[NOTE]
====
Successful data operations confirm volume provisioning and mounting functionality. However, this verification alone does not validate encryption. The following steps provide cryptographic verification.
====

=== Step 6: Cryptographic verification

To conclusively verify that encryption is active, direct inspection of the underlying storage device is required.

==== 6.1. Identify the volume location

Determine which node hosts the volume replica:

[source,bash]
----
# Get the PVC details
kubectl get pvc test-encrypted-pvc

# Get the associated PersistentVolume
kubectl get pv | grep test-encrypted

# View Longhorn volume details
kubectl get volumes -n longhorn-system | grep test-encrypted
----

==== 6.2. Access the node and locate volume files

SSH to the node hosting the volume (identified from the previous output):

[source,bash]
----
# Connect to the node
ssh root@<node-ip>

# Locate the volume image file
find /var/lib/longhorn -name "*.img*" -type f | grep -i test
----

Typical volume path:
[source,text]
----
/var/lib/longhorn/replicas/pvc-abc12345-r-6789/volume-head-001.img
----

==== 6.3. Attempt plaintext data extraction

> Execute the following command to search for unencrypted data:

Reviewer: ```suggestion
Run the following command to search for unencrypted data:
```

=== Step 8: Data persistence verification

Verify that encrypted volumes maintain data integrity across pod lifecycle events:

[source,bash]
----
# Delete the pod (volume persists)
kubectl delete pod encryption-test-pod

# Recreate the pod
kubectl apply -f test-encryption.yaml

# Wait for Running state
kubectl get pod encryption-test-pod -w

# Verify data persistence
kubectl exec -it encryption-test-pod -- cat /data/secret.txt
----

Data should remain intact, demonstrating proper volume persistence.

== Encryption verification summary

The following table summarizes encryption verification indicators:

|===
|Verification Method | Encrypted | Not Encrypted

|`strings` output
|No plaintext found
|Plaintext strings visible

|`hexdump` output
|Random binary data
|Readable ASCII text

|`lsblk` output
|`crypto_LUKS` type present
|No `crypto_LUKS` type

|`dmsetup ls --target crypt`
|Lists crypt devices
|No crypt devices listed
|===

== Test environment cleanup

Remove test resources:

[source,bash]
----
# Delete the test pod
kubectl delete -f test-encryption.yaml

# Delete the PVC and volume
kubectl delete pvc test-encrypted-pvc
----

== CAPI integration

> When deploying clusters using Cluster API (CAPI) with Metal3, encryption configuration can be embedded in cluster provisioning manifests, enabling encrypted storage from initial cluster bootstrap.

Reviewer: ```suggestion
When deploying clusters using Cluster API (CAPI) with Metal^3^, encryption configuration can be embedded in cluster provisioning manifests, enabling encrypted storage from initial cluster bootstrap.
```

+
[source,bash]
----
openssl rand -base64 32
----
+
Example output:
+
[source,text]
----
Kq8Xh2vN9pL3mZ1wR4tY6uI0oP5aS7dF8gH9jK1lM2n=
----
+
[WARNING]
====
**Key Management Critical Requirements**

* **Never commit encryption keys to version control systems**
* Loss of the encryption key results in **permanent and irreversible data loss**
====

. **Create the encryption key Secret**
+
Once generated, the encryption key must be stored as a Kubernetes Secret in the `longhorn-system` namespace:
+
[source,bash]
----
CRYPTO_KEY_VALUE="Kq8Xh2vN9pL3mZ1wR4tY6uI0oP5aS7dF8gH9jK1lM2n="

kubectl create secret generic longhorn-crypto \
  --from-literal=CRYPTO_KEY_VALUE="${CRYPTO_KEY_VALUE}" \
  -n longhorn-system
----
+
Verify successful Secret creation:
+
[source,bash]
----
kubectl get secret longhorn-crypto -n longhorn-system
----
+
[TIP]
====
In production environments, consider using external secret management solutions that integrate with Kubernetes, such as the External Secrets Operator or sealed-secrets, to avoid storing sensitive data directly in the cluster's etcd.
====

. **Configure the encrypted StorageClass**
+

> Define a `StorageClass` that specifies encryption parameters for the Longhorn CSI driver:

Reviewer: ```suggestion
. **Configure the encrypted StorageClass**

.. Define a `StorageClass` that specifies encryption parameters for the Longhorn CSI driver:
```

metadata:
  name: longhorn-encrypted
provisioner: driver.longhorn.io
allowVolumeExpansion: true
parameters:
  numberOfReplicas: "3"
  staleReplicaTimeout: "2880"
  fromBackup: ""
  fsType: "ext4"
  dataLocality: "disabled"
  # Encryption configuration
  encrypted: "true"
  csi.storage.k8s.io/provisioner-secret-name: "longhorn-crypto"
  csi.storage.k8s.io/provisioner-secret-namespace: "longhorn-system"
  csi.storage.k8s.io/node-publish-secret-name: "longhorn-crypto"
  csi.storage.k8s.io/node-publish-secret-namespace: "longhorn-system"
  csi.storage.k8s.io/node-stage-secret-name: "longhorn-crypto"
  csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
----
+
The encryption-specific parameters configure CSI secret references:
+
* `encrypted: "true"` - Enables LUKS encryption for all volumes provisioned from this StorageClass
* `csi.storage.k8s.io/provisioner-secret-name` - Specifies the Secret containing the encryption key used during volume provisioning
* `csi.storage.k8s.io/node-publish-secret-name` - Identifies the Secret used when mounting volumes to pods
* `csi.storage.k8s.io/node-stage-secret-name` - References the Secret used during volume staging operations
+

> Apply the StorageClass definition:

Reviewer: ```suggestion

.. Apply the StorageClass definition:
```

+
[source,bash]
----
kubectl apply -f longhorn-encrypted-storageclass.yaml
----
+

> Verify StorageClass creation and configuration:

Reviewer: ```suggestion

. Verify StorageClass creation and configuration:
```

+
[source,bash]
----
kubectl get storageclass longhorn-encrypted
kubectl get storageclass longhorn-encrypted -o yaml
----

. **Deploy a test workload**
+
To validate the encryption implementation, deploy a test pod that writes data to an encrypted volume.
+

> Create the following manifest (`test-encryption.yaml`):

Reviewer: ```suggestion
To validate the encryption implementation, deploy a test pod that writes data to an encrypted volume.

.. Create the following manifest (`test-encryption.yaml`):
```

  resources:
    requests:
      storage: 1Gi

---
# Pod that writes sensitive data to the volume
apiVersion: v1
kind: Pod
metadata:
  name: encryption-test-pod
  namespace: default
  labels:
    app: encryption-test
spec:
  containers:
  - name: writer
    image: busybox
    command:
      - sh
      - -c
      - |
        echo "=== Writing sensitive data to encrypted volume ==="
        echo "SECRET DATA: Credit Card 4111-1111-1111-1111" > /data/secret.txt
        echo "SECRET DATA: Password MySuperSecret123!" >> /data/secret.txt
        echo "SECRET DATA: API Key sk-1234567890abcdef" >> /data/secret.txt
        date >> /data/secret.txt
        echo "=== Data written successfully ==="
        echo "Content of /data/secret.txt:"
        cat /data/secret.txt
        echo "=== Sleeping forever (use 'kubectl exec' to verify) ==="
        sleep infinity
    volumeMounts:
    - name: encrypted-storage
      mountPath: /data
  volumes:
  - name: encrypted-storage
    persistentVolumeClaim:
      claimName: test-encrypted-pvc
----
+
This manifest defines:
+
. A `PersistentVolumeClaim` requesting 1Gi of storage from the `longhorn-encrypted` StorageClass

> . A Pod that writes test data to the encrypted volume and remains running for verification

Reviewer: ```suggestion
+
* A `PersistentVolumeClaim` requesting 1Gi of storage from the `longhorn-encrypted` StorageClass
* A Pod that writes test data to the encrypted volume and remains running for verification
```

+

> Deploy the test workload:

Reviewer: ```suggestion

.. Deploy the test workload:
```

+
[source,bash]
----
kubectl apply -f test-encryption.yaml
----
+

> Monitor pod startup:

Reviewer: ```suggestion

.. Monitor pod startup:
```

kubectl delete pod encryption-test-pod

# Recreate the pod
kubectl apply -f test-encryption.yaml

# Wait for Running state
kubectl get pod encryption-test-pod -w

# Verify data persistence
kubectl exec -it encryption-test-pod -- cat /data/secret.txt
----
+
Data should remain intact, demonstrating proper volume persistence.

== Encryption verification summary

The following table summarizes encryption verification indicators:

|===
|Verification Method | Encrypted | Not Encrypted

|`strings` output
|No plaintext found
|Plaintext strings visible

|`hexdump` output
|Random binary data
|Readable ASCII text

|`lsblk` output
|`crypto_LUKS` type present
|No `crypto_LUKS` type

|`dmsetup ls --target crypt`
|Lists crypt devices
|No crypt devices listed
|===

== Test environment cleanup

Remove test resources:

[source,bash]
----
# Delete the test pod
kubectl delete -f test-encryption.yaml

# Delete the PVC and volume
kubectl delete pvc test-encrypted-pvc
----

== CAPI integration

When deploying clusters using Cluster API (CAPI) with Metal^3^, encryption configuration can be embedded in cluster provisioning manifests, enabling encrypted storage from initial cluster bootstrap.

> A complete reference implementation is available in the https://github.com/suse-edge/telco-cloud-examples/tree/main/telco-examples/downstream-clusters/longhorn/longhorn-with-encryption[SUSE Edge telco-cloud-examples repository].

Reviewer: The external link leads me to 404 page. Can you please check?

Author: need to merge the PR in the other repo referenced here, but the url should work after that
